lib/ABI/FunctionType/Layout.cpp

```diff
@@ -0,0 +1,963 @@
+  /// \param PassesReturnValueLocationAsAnArgument `true` if the first argument
+  ///        slot should be occupied by a shadow return value, `false` otherwise
+  /// \return Information about registers and stack that are to be used to
+  ///         pass said arguments.
+  DistributedArguments
+  distributeArguments(const ArgumentSet &Arguments,
+                      bool PassesReturnValueLocationAsAnArgument) const;
+
+  /// Helper for choosing the "Raw" type of a specific argument or return
+  /// value based on its "C" type and the register it is to be placed in.
+  ///
+  /// \param ArgumentType The original "C" type.
+  /// \param Register Register this value is placed in.
+  /// \param Bucket The bucket to store the type in if it's necessary to add
+  ///        a new one.
+  ///
+  /// \return The resulting "Raw" type for the current entity.
+  model::QualifiedType chooseType(const model::QualifiedType &ArgumentType,
+                                  const model::Register::Values Register,
+                                  TypeBucket &Bucket) const;
+
+private:
+  DistributedArguments
+  distributePositionBasedArguments(const ArgumentSet &Arguments,
+                                   std::size_t SkippedCount = 0) const;
+  DistributedArguments
+  distributeNonPositionBasedArguments(const ArgumentSet &Arguments,
+                                      std::size_t SkippedCount = 0) const;
+
+  std::pair<DistributedArguments, std::size_t>
+  distributeImpl(model::QualifiedType Type,
+                 RegisterSpan Registers,
+                 std::size_t OccupiedRegisterCount,
+                 std::size_t AllowedRegisterLimit,
+                 bool AllowPuttingPartOfAnArgumentOnStack) const;
+
+public:
+  uint64_t finalStackOffset(const DistributedArguments &Arguments) const;
+};
+
+model::QualifiedType
+ToRawConverter::chooseType(const model::QualifiedType &ArgumentType,
+                           model::Register::Values Register,
+                           TypeBucket &Bucket) const {
+  auto MaybeSize = ArgumentType.size();
+  auto TargetSize = model::Register::getSize(Register);
+
+  if (!MaybeSize.has_value()) {
+    return model::QualifiedType(Bucket.defaultRegisterType(Register), {});
+  } else if (*MaybeSize > TargetSize) {
+    // TODO: additional considerations should be made here
+    //       for better position-based ABI support.
+    const auto Architecture = model::ABI::getArchitecture(ABI.ABI());
+    return ArgumentType.getPointerTo(Architecture);
+  } else if (!ArgumentType.isScalar()) {
+    // TODO: this can be considerably improved, we can preserve more
+    //       information here over just returning the default type.
+    return model::QualifiedType(Bucket.defaultRegisterType(Register), {});
+  }
+
+  return ArgumentType;
+}
+
+model::TypePath
+ToRawConverter::convert(const model::CABIFunctionType &Function,
+                        TupleTree<model::Binary> &Binary) const {
```

Comment:
The use cases for this conversion path are basically only for testing, this function needs to be pruned to be as simple as possible so that we can test what we want to test.
Preserving names, struct fields and the like is out of scope and a burden to maintain.

Reply:
What do you think about it now? (`ToRawConverter::convert` in `Layout.cpp`)

Reply:
Seems good!